Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm

```diff
@@ -296,6 +357,187 @@ - (void)_loadMetal
     return [[_WKARPresentationSession alloc] initWithSession:session descriptor:descriptor];
 }
 
+#pragma mark - _WKTransientAction
+
+@interface _WKTransientAction : NSObject
+@property (nonatomic, readonly) simd_float4x4 targetRay;
+@property (nonatomic, readonly) simd_float4x4 startingPoseInverse;
+@property (nonatomic) simd_float4x4 pose;
+- (nullable instancetype)initWithTargetRay:(simd_float4x4)targetRay pose:(simd_float4x4)pose;
+@end
+
+@implementation _WKTransientAction
+
+- (nullable instancetype)initWithTargetRay:(simd_float4x4)targetRay pose:(simd_float4x4)pose
+{
+    self = [super init];
+    if (self) {
+        _targetRay = targetRay;
+        _startingPoseInverse = simd_inverse(pose);
+        _pose = pose;
+    }
+    return self;
+}
+
+@end
+
+#pragma mark - _WKTransientGestureRecognizer implementation
+
+@implementation _WKTransientGestureRecognizer {
+    OSObjectPtr<dispatch_queue_t> _accessQueue;
+    WeakObjCPtr<_WKARPresentationSession> _session;
+    RetainPtr<NSMutableDictionary<NSNumber *, _WKTransientAction *>> _transientActions;
+}
+
+- (nullable instancetype)initWithSession:(_WKARPresentationSession *)session
+{
+    self = [super init];
+    if (self) {
+        _accessQueue = adoptOSObject(dispatch_queue_create("com.apple.WebContent._WKTransientGestureRecognizer.AccessQueue", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL));
+        _session = session;
+        _transientActions = adoptNS([NSMutableDictionary new]);
+    }
+    return self;
+}
+
+- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
+{
+    if (!_session)
+        return;
+
+    ARFrame* frame = [_session currentFrame];
+    if (!frame)
+        return;
+
+    // We differ from visionOS here and use the ARCamera as the pose reference
+    // space.
+    simd_float4x4 poseTransform = frame.camera.transform;
+
+    UIView* view = [_session view];
+    for (UITouch *touch in touches) {
+        auto normalizedPoint = [touch normalizedLocationInView:view];
+        auto targetRayTransform = [_session raycastQueryTransformFromPoint:normalizedPoint];
+
+        dispatch_async(_accessQueue.get(), ^{
+            NSNumber *id = @(touch.hash);
+            RetainPtr<_WKTransientAction> transientAction = adoptNS([[_WKTransientAction alloc] initWithTargetRay:targetRayTransform pose:poseTransform]);
+            [_transientActions setObject:transientAction.get() forKey:id];
+            RELEASE_LOG_DEBUG(XR, "Transient action started: %@", id);
+        });
+    }
+}
+
+- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
+{
+    RELEASE_LOG(XR, "%s", __FUNCTION__);
+
+    if (!_session)
+        return;
+
+    ARFrame* frame = [_session currentFrame];
```

Comment:
Nit: * on the other side.

Reply:
RetainPtr

Reply:
From @cdumez comment, I take it we should never have a raw ObjC ptr and let ARC deal with it?

Reply:
If we're using ARC, then raw pointers are fine. If not, then our modern smart pointer rules apply.
I don't know if we're using ARC here. In my experience, ARC is disabled for most of WebKit.